llvm/lib/Target/RISCV/RISCVInsertVSETVLI.cpp

```diff
@@ -519,28 +527,31 @@ class VSETVLIInfo {
   bool getTailAgnostic() const { return TailAgnostic; }
   bool getMaskAgnostic() const { return MaskAgnostic; }
 
-  bool hasNonZeroAVL() const {
+  bool hasNonZeroAVL(const LiveIntervals *LIS) const {
     if (hasAVLImm())
       return getAVLImm() > 0;
-    if (hasAVLReg())
-      return isNonZeroLoadImmediate(getAVLDefMI());
+    if (hasAVLReg()) {
+      if (auto *DefMI = LIS->getInstructionFromIndex(getAVLVNInfo()->def))
+        return isNonZeroLoadImmediate(*DefMI);
+    }
     if (hasAVLVLMAX())
       return true;
     if (hasAVLIgnored())
       return false;
     return false;
   }
 
-  bool hasEquallyZeroAVL(const VSETVLIInfo &Other) const {
+  bool hasEquallyZeroAVL(const VSETVLIInfo &Other,
+                         const LiveIntervals *LIS) const {
     if (hasSameAVL(Other))
       return true;
-    return (hasNonZeroAVL() && Other.hasNonZeroAVL());
+    return (hasNonZeroAVL(LIS) && Other.hasNonZeroAVL(LIS));
   }
 
   bool hasSameAVL(const VSETVLIInfo &Other) const {
     if (hasAVLReg() && Other.hasAVLReg())
-      return AVLRegDef.DefMI == Other.AVLRegDef.DefMI &&
-             AVLRegDef.DefReg == Other.AVLRegDef.DefReg;
+      return getAVLVNInfo()->id == Other.getAVLVNInfo()->id &&
```

Comment:
Maybe `getAVLVNInfo()->def == Other.getAVLVNInfo()->def` is better than id?

Reply:
I didn't notice any difference between using id and def, but is it possible for two different phi nodes to end up with the same def after phi elimination? I.e. both would point to the start of the block

Reply:
> I didn't notice any difference between using id and def, but is it possible for two different phi nodes to end up with the same def after phi elimination? I.e. both would point to the start of the block

When VNInfo [be created](https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/CodeGen/LiveInterval.h#L333), the id is `valnos.size()` inside `LiveRange`. it could be same if they are different LiveRange.

I think it safe to use `id` because `getAVLReg() == Other.getAVLReg()` make sure they are same LiveInterval.